src/lib/reasoners/arrays_rel.ml

```diff
@@ -166,25 +166,59 @@ module Debug = struct
 end
 (*BISECT-IGNORE-END*)
 
-(* met a jour gets et tbset en utilisant l'ensemble des termes donne*)
-let rec update_gets_sets acc t =
+let merge_revelant_terms (gets, tbset) (g, t) =
+  let gets = G.union gets g in
+  let tbset =
+    TBS.merge
+      (fun _ s1 s2 ->
+         match s1, s2 with
+         | Some s1, Some s2 -> Some (S.union s1 s2)
+         | Some s, None | None, Some s -> Some s
+         | None, None -> None
+      ) tbset t
+  in
+  gets, tbset
+
+(* Collects all the select or store subterms of the term [t]
+   for the instantiation engine. Use a weak cache to avoid a bottleneck in
+   presence of very large terms in the problem.
+
+   See issue https://github.com/OCamlPro/alt-ergo/issues/1123 *)
+let rec relevant_terms t =
   let { E.f; xs; ty; _ } = E.term_view t in
-  let gets, tbset = List.fold_left update_gets_sets acc xs in
+  let gets, tbset =
+    List.fold_left
+      (fun acc x ->
+         merge_revelant_terms acc (cached_relevant_terms x)
+      ) (G.empty, TBS.empty) xs
+  in
   match Sy.is_get f, Sy.is_set f, xs with
   | true , false, [a;i]   -> G.add {g=t; gt=a; gi=i; gty=ty} gets, tbset
   | false, true , [a;i;v] ->
     gets, TBS.add a {s=t; st=a; si=i; sv=v; sty=ty} tbset
   | false, false, _ -> (gets,tbset)
   | _  -> assert false
 
+and cached_relevant_terms =
+  let module H = Ephemeron.K1.Make (Expr) in
+  let cache : (G.t * S.t TBS.t) H.t = H.create 1024
```

Comment:
I think this breaks Alt-Ergo-Fuzz and makes @hra687261 sad

Reply:
Thanks for keeping it in mind!
I suppose it can be kept in the environment or it can expose a cache resetting function which should be called in the right places.
But the first solution seems cleaner. I would assume that performance differences would be negligible.

Reply:
@hra687261 Out of sheer curiosity, why is it important for a fuzzing tool?

Reply:
It uses Alt-Ergo as a library, and to make sure that two independent runs have the same behavior, it is necessary to be able to reset and the resolution context (hash-tables, refs, environment). Otherwise the fuzzing doesn't work. (cf. https://github.com/OCamlPro/alt-ergo/pull/490)